### PR TITLE
REL: filter out @ in authors name and add count

### LINKED ISSUE #15543: REL: list contributors using GitHub handles
When we do the release on GitHub, I guess we copy/paste the text we put in the rst files. If yes, then can we adjust the contributor list to use the proper GitHub handles? Because right now GitHub only list/link a few contributors instead of the >100 contributors who effectively contributed. It would help the recognition of smaller contributors.

As we can see in the screenshot, only if you use the GH handle, the UI will pick it up.

Screenshot 2022-02-07 at 13 22 20

### PR DESCRIPTION
**Pull request by tupui**
Closes #15543

Does the following:

* remove `@` if the name is being mapped to GitHub handle
* remove `GitHub` user automatically
* take into account co-authors
* add a count of commit per user (can remove if you don't like it, can also order by count instead of names, could also not show count when it's just 1 commit)
* (wrong title level while browsing the release page in the doc)

See bellow some example:

* `@` removed: `python tools/authors.py 9a504cd27ad..9cbbbba944`

```
# before this PR

* @endolith
* Andrew Annex +
* Nickolai Belakovski
* GitHub
* Xingyu Liu
* Cong Ma
* Tyler Reddy
* Leo C. Stein +
* Kai Striega
* Warren Weckesser

A total of 10 people contributed to this release.
```

```
# this PR

* endolith (1)
* Andrew Annex (1) +
* Nickolai Belakovski (1)
* Xingyu Liu (4)
* Cong Ma (2)
* Tyler Reddy (4)
* Leo C. Stein (4) +
* Kai Striega (18)
* Warren Weckesser (4)

A total of 9 people contributed to this release.
```

* `Co-authored-by:` accounted: `python tools/authors.py e7f5741df..b06254e5a08a`

```
# before this PR

* Gerrit Ansmann
* GitHub
* Ralf Gommers
* P. L. Lim

A total of 4 people contributed to this release.
```

```
# this PR

* Gerrit Ansmann (1)
* Ralf Gommers (1)
* P. L. Lim (1)
* Pamphile Roy (1)
* Albert Steppi (1)

A total of 5 people contributed to this release.
```

### CODE REVIEW

File: tools/authors.py
Comment (rgommers) [inline tools/authors.py:8]: Better leave this out, since it's typically the tag rather than a commit hash. This is just the standard Git way of specifying a commit range.

File: tools/authors.py
Comment (rgommers) [inline tools/authors.py:65]: good catch, this is a helpful change

**Review by rgommers — APPROVED** [review]
Thanks @tupui. LGTM, modulo my one small doc remark. I'll leave it to @tylerjereddy to merge.

**Review by tylerjereddy — APPROVED** [review]
Looks pretty good--I did `git diff --no-index --word-diff authors_draft_orig.txt authors_draft_new.txt` locally to compare this branch against `main` for a command I did in the `1.8.x` series (`python tools/authors.py 8b5e5c8cd58..169749b >& authors_draft.txt`), and all of the changes were improvements (see below the fold).

Not sure on counting commits, but don't really mind.

<details>

```diff
--- a/authors_draft_orig.txt
+++ b/authors_draft_new.txt
@@ -2,146 +2,150 @@
Authors
=======

* [-@endolith-]{+99991 (1) ++}
{+* endolith (5)+}
* adamadanandy {+(2)+} +
* akeemlh {+(1)+} +
* Anton Akhmerov {+(2)+}
* Marvin Albert {+(1)+} +
* alegresor {+(1)+} +
* Andrew Annex {+(1)+} +
* Pantelis Antonoudiou {+(2)+} +
* Ross Barnowski {+(41)+} +
* Christoph Baumgarten {+(7)+}
* Stephen Becker {+(1)+} +
* Nickolai Belakovski {+(1)+}
* Peter Bell {+(49)+}
* berberto {+(2)+} +
* Georgii Bocharov {+(1)+} +
* Evgeni Burovski {+(25)+}
* Matthias Bussonnier {+(3)+}
* CJ Carey {+(2)+}
* Justin Charlong {+(2)+} +
* Hood Chatham {+(2)+} +
* Dennis Collaris {+(1)+} +
* David Cottrell {+(2)+} +
* cruyffturn {+(1)+} +
* da-woods {+(1)+} +
* Anirudh Dagar {+(44)+}
* Tiger Du {+(1)+} +
* Thomas Duvernay {+(6)+}
* Dani El-Ayyass {+(2)+} +
* Castedo Ellerman {+(1)+} +
* Donnie Erb {+(2)+} +
* Andreas Esders-Kopecky {+(1)+} +
* Livio F {+(1)+} +
* Isuru Fernando {+(9)+}
* Evelyn Fitzgerald {+(1)+} +
* Sara Fridovich-Keil {+(2)+} +
* Mark E Fuller {+(2)+} +
*[-GitHub-]
[-*-] Ralf Gommers {+(216)+}
* Kevin Richard Green {+(1)+} +
* guiweber {+(1)+} +
* Nitish Gupta {+(1)+} +
* h-vetinari {+(1)+}
* Matt Haberland {+(221)+}
* J. Hariharan {+(1)+} +
* Charles Harris {+(1)+}
* Jonathan Helgert {+(1)+} +
* Trever Hines {+(1)+}
* Ian Hunt-Isaak {+(1)+} +
* ich {+(2)+} +
* {+Kei Ishikawa (1)+}
{+*+} Itrimel {+(1)+} +
* Jan-Hendrik Müller {+(1) ++}
{+* jarneal (1)+} +
* Jebby993 {+(1)+} +
* Yikun Jiang {+(1)+} +
* Evan W Jones {+(1)+} +
* Nathaniel Jones {+(1)+} +
* Jeffrey Kelling {+(4)+} +
* Malik Idrees Hasan Khan {+(2)+} +
* {+Paul Kienzle (1)+}
{+*+} Sergey B Kirpichev {+(1)+}
* Kadatatlu Kishore {+(1)+} +
* Andrew Knyazev {+(2)+}
* Ravin Kumar {+(1)+} +
* Peter Mahler Larsen {+(6)+}
* Eric Larson {+(12)+}
* Antony Lee {+(3)+}
* Gregory R. Lee {+(1)+}
* Tim Leslie {+(2)+}
* lezcano {+(2)+} +
* Xingyu Liu {+(45)+}
* Christian Lorentzen {+(2)+}
* Lorenzo {+(2)+} +
* Smit Lunagariya {+(121)+} +
* Lv101Magikarp {+(1)+} +
* Yair M {+(1)+} +
* Cong Ma {+(2)+}
* Lorenzo Maffioli {+(2)+} +
* majiang {+(1)+} +
* Brian McFee {+(1)+} +
* Nicholas McKibben {+(8)+}
* John Speed Meyers {+(1)+} +
* millivolt9 {+(1)+} +
* Jarrod Millman {+(2)+}
* Harsh Mishra {+(2)+} +
* Boaz Mohar {+(1)+} +
* naelsondouglas {+(1)+} +
* Andrew Nelson {+(94)+}
* Nico Schlömer {+(43)+}
* Thomas Nowotny {+(1)+} +
* nullptr {+(1)+} +
* Teddy Ort {+(3)+} +
* Nick Papior {+(4)+}
* ParticularMiner {+(1)+} +
* Dima Pasechnik {+(5)+}
* Tirth Patel {+(42)+}
* Matti Picus {+(2)+}
* Ilhan Polat {+(51)+}
* Adrian Price-Whelan {+(2)+} +
* Quentin Barthélemy {+(4)+} +
* Sundar R {+(1)+} +
* Judah Rand {+(4)+} +
* Tyler Reddy {+(152)+}
* Renal-Of-Loon {+(1)+} +
* Frederic Renner {+(2)+} +
* Pamphile Roy {+(93)+}
* Bharath Saiguhan {+(1)+} +
* Atsushi Sakai {+(37)+}
* Eric Schanet {+(1)+} +
* Sebastian Wallkötter {+(3)+}
* serge-sans-paille {+(6)+}
* Reshama Shaikh {+(3)+} +
* Namami Shanker {+(2)+}
* siddhantwahal {+(1)+} +
* Walter Simson {+(1)+} +
* Gagandeep Singh {+(38)+} +
* Leo C. Stein {+(4)+} +
* Albert Steppi {+(4)+}
* Kai Striega {+(19)+}
* Diana Sukhoverkhova {+(1)+}
* Søren Fuglede Jørgensen {+(2)+}
* Masayuki Takagi {+(1)+} +
* Mike Taves {+(3)+}
* Ben Thompson {+(2)+} +
* Bas van Beek {+(47)+}
* Jacob Vanderplas {+(2)+}
* Dhruv Vats {+(2)+} +
* H. Vetinari {+(25)+} +
* Thomas Viehmann {+(1)+} +
* Pauli Virtanen {+(2)+}
* Vlad {+(1)+} +
* Arthur Volant {+(23)+}
* Samuel Wallan {+(2)+}
* Stefan van der Walt {+(36)+}
* Warren Weckesser {+(39)+}
{+* Michael Wende (1) ++}
* Josh Wilson {+(1)+}
* Haoyin Xu {+(4)+} +
* Rory Yorke {+(2)+}
* Egor Zemlyanoy {+(2)+}
* Gang Zhao {+(15)+} +
* 赵丰 (Zhao Feng) {+(1)+} +

A total of [-138-]{+142+} people contributed to this release.
People with a "+" by their names contributed a patch for the first time.
This list of names is automatically generated, and may not be fully complete.

```

</details>

File: tools/authors.py
Comment (tylerjereddy) [reply tools/authors.py:8]: Looks like I mostly use hashes for the authors script and tags for the list script so at least for me I don't mind either way.

**Comment by tylerjereddy** [comment]
I'll merge this in thanks, and maybe backport too.

Some other thoughts:
- the update here reveals more opportunities to update `.mailmap` of course, because now new co-authors show up (I see one that is digits-only like `99991` or something like that)
- the `scikit-learn` example cited is nice, not sure how they do that; not sure if we'd want to invest the time to do it or not
- some other annoying things that happen with release notes:
  - when I paste the log I have to manually delete stuff like `-------- -` in headers and `- -` in bullet lists because of encryption apparently (see old mailing list exchange b/w Pauli and I), and I don't think we really use the encryption stuff anymore; at some point we should probably decide if we want to keep that encryption stuff, and if there is a way to do it without me re-formatting the headings/bullet lists for the stray `-` characters, which does get a bit old for both the email and the GitHub release page *every time*

**Review by tupui — COMMENTED** [review]
Thank you both for having a look at this.

> I'll merge this in thanks, and maybe backport too.
>
> Some other thoughts:
>
> the update here reveals more opportunities to update .mailmap of course, because now new co-authors show up (I see one that is digits-only like 99991 or something like that)
the scikit-learn example cited is nice, not sure how they do that; not sure if we'd want to invest the time to do it or not
some other annoying things that happen with release notes:
when I paste the log I have to manually delete stuff like -------- - in headers and - - in bullet lists because of encryption apparently (see old mailing list exchange b/w Pauli and I), and I don't think we really use the encryption stuff anymore; at some point we should probably decide if we want to keep that encryption stuff, and if there is a way to do it without me re-formatting the headings/bullet lists for the stray - characters, which does get a bit old for both the email and the GitHub release page every time

I would be happy to help you solve these issues if you can point me/create an issue so that I have a full spec of what would really make sense for you. There are tons of moving part during the release process for you to handle so if I can help to remove small things feel free to ping me 😃

**Pull request merged**
